ramips: improve MT7621S core detection

namiltd opened the pull request:
The proposed detection method was based on reading the LAUNCH_FREADY core flag.
However, this method only works before the cores are launched.
For this reason, the core number detection method has been changed to a simpler one.
For mt6721s the 17th revision bit is zero, hence we know that it is this chip,
so the number of cores is 1.

Fixes: https://github.com/openwrt/openwrt/issues/17764

Tested on:
- Dual core routers (MT7621A/MT7621DA):
  -  DNA Valokuitu Plus EX400
  - Maginon MC-1200AC
- Single core routers (MT7621S):
  -  Netgear R6220

namiltd commented:
The fix should be applied to main and openwrt-24.10 branches.

namiltd commented:
Waiting for more tests.

DragonBluep commented on target/linux/ramips/patches-6.6/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch:
```diff
@@ -56,16 +58,20 @@ Signed-off-by: Thomas Bogendoerfer <tsbogend@alpha.franken.de>
++		launch++; /* core0:1 */
++		flags01 = launch->flags;
++		launch++; /* core1:0 */
++		if ((flags01 & LAUNCH_FREADY) && !(launch->flags & LAUNCH_FREADY))
```
It's not hard to detect the MT7621s variant by reading the syscon register.
```
	#include <asm/mach-ralink/mt7621.h>
	#define CHIP_REV_CORE_MASK		0x1
	#define CHIP_REV_CORE_SHIFT		17
	if (IS_ENABLED(CONFIG_SOC_MT7621)) {
		u32 sysc = __raw_readl(MT7621_SYSC_BASE + SYSC_REG_CHIP_REV);
		if (!((sysc >> CHIP_REV_CORE_SHIFT) & CHIP_REV_CORE_MASK))
			ncores = 1;
	}
```

namiltd replied:
CPU0 revision is: 0001992f for single and dual core version.

DragonBluep replied:
> CPU0 revision is: 0001992f for single and dual core version.

It's mips cpu revision, not soc revision. MTK SDK uses this way to start different mt7621 variants.
https://github.com/search?q=repo%3Agl-inet%2Fuboot-source-for_mtk%20RALINK_DUAL_CORE_FUN%20&type=code

Rondom replied:
https://source.denx.de/u-boot/u-boot/-/blob/2b1c8d3b2da46ce0f7108f279f04bc66f1d8d09a/arch/mips/mach-mtmips/mt7621/init.c#L28-53

DragonBluep replied:
@mrkiko Why you check `SYSC_REG_CHIP_NAME0` register instead of `SYSC_REG_CHIP_REV`?

mrkiko commented:
To add some context:
landing this fix upstream (Torvalds tree) was attempted - it was applied via commit 6decd1aad15f56b169217789630a0098b496de0e and then reverted via commit 740da9d7ca4e25f5d87db9f80d75432681b61305 .

mrkiko commented:
I tried the approach - and it returns 1 as expected.
Pasting here the hacked up function just for reference...

Note that I added
#include <asm/mach-ralink/mt7621.h>
#define CPU_ID 0x20000

ad the beginning...

static inline unsigned int mips_cps_numcores(unsigned int cluster)
{
	unsigned int ncores, testcores, testval;

	if (!mips_cm_present())
		return 0;

	/* Add one before masking to handle 0xff indicating no cores */
	ncores = FIELD_GET(CM_GCR_CONFIG_PCORES,
			 mips_cps_cluster_config(cluster) + 1);

	if (IS_ENABLED(CONFIG_SOC_MT7621)) {
		struct cpulaunch *launch;
		unsigned long flags01, flags10;

		/*
		 * Ralink MT7621S SoC is single core, but the GCR_CONFIG method
		 * always reports 2 cores. Check the second core's LAUNCH_FREADY
		 * flag to detect if the second core is missing. This method
		 * only works before the core has been started.
		 */
		launch = (struct cpulaunch *)CKSEG0ADDR(CPULAUNCH);
		pr_info("cpu0:0 flags = %lu, ", launch->flags); launch += 1; flags01 = launch->flags; pr_info("cpu0:0 flags = %lu, ", flags01);
		launch += 1; /* MT7621 has 2 VPEs per core */ flags10 = launch->flags; pr_info("cpu1:0 flags = %lu, ", flags10); launch += 1; pr_info("cpu1:1 flags = %lu\n", launch->flags);
		if ((flags01 & LAUNCH_FREADY) && !(flags10 & LAUNCH_FREADY))
			ncores = 1;

		testval = __raw_readl(MT7621_SYSC_BASE + SYSC_REG_CHIP_NAME0);
		testcores = FIELD_GET(CPU_ID, testval);
		pr_info("Using readl, it looks like my system has %u cores.\n",testcores);
	}

	return ncores;
}


R6220 says:
Using readl, it looks like my system has 1 cores.

On Mon, Feb 03, 2025 at 06:54:08PM -0800, Andreas Gnau wrote:
> @Rondom commented on this pull request.
> 
> 
> 
> >  +		 */
>  +		launch = (struct cpulaunch *)CKSEG0ADDR(CPULAUNCH);
> -+		launch += 2; /* MT7621 has 2 VPEs per core */
> -+		if (!(launch->flags & LAUNCH_FREADY))
> ++		launch++; /* core0:1 */
> ++		flags01 = launch->flags;
> ++		launch++; /* core1:0 */
> ++		if ((flags01 & LAUNCH_FREADY) && !(launch->flags & LAUNCH_FREADY))
> 
> https://source.denx.de/u-boot/u-boot/-/blob/2b1c8d3b2da46ce0f7108f279f04bc66f1d8d09a/arch/mips/mach-mtmips/mt7621/init.c#L28-53
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/17834#discussion_r1940409237
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

mrkiko commented:
I am not stating this is the perfect solution - and infact would be interested in understanding why a different path has been chosen.
@nbd168 @dangowrt I am pretty sure you know better. :)

namiltd commented:
`testval = _mt7621_get_soc_name0;`

mrkiko commented:
Modifying the pr_info call to look like:
pr_info("Using readl, it looks like my system has %u cores (testval = %u).\n",testcores,testval);
(being the two values unsigned integers, getting no warnings)

results in:
[    0.169103] Using readl, it looks like my system has 1 cores (testval = 909595725).

On Tue, Feb 04, 2025 at 02:27:08PM -0800, Mieczyslaw Nalewaj wrote:
> ```
> 		testval = _mt7621_get_soc_name0();
> 		testcores = FIELD_GET(CPU_ID, testval);
> 		pr_info("Using readl, it looks like my system has %u cores.\n",testcores)
> ```
> According to arch/mips/ralink/mt7621.c:
> On mt7621 mt7621_get_soc_name0() should be equal to MT7621_CHIP_NAME0 (0x3637544D)
> 0x3637544D & 0x20000 = 0x20000
> And that would mean that on mt7621 testcore is always equal to 2
> 
> @mrkiko Could you print the value of testval?
> 
> 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/17834#issuecomment-2635134504
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

mrkiko commented:
So we have our answer now. That said, I plan to keep the R6220 running for some more days. Feel free to contact me if you need me to test any code.

namiltd commented:
The code from this PR works best, but it still needs testing on other devices with mt7621s.

mrkiko commented:
Hello!! I think I have messed up.

Modifying the code to look like:
    testval = __raw_readl(MT7621_SYSC_BASE + 0x0c);
    testcores = FIELD_GET(CPU_ID, testval);
    pr_info("Using readl, it looks like my system has %u cores (testval = %u).\n",testcores,testval);

I get:

[    0.169101] Using readl, it looks like my system has 0 cores (testval = 65795).


On Fri, Feb 07, 2025 at 06:08:12AM -0800, DragonBluep wrote:
> @DragonBluep commented on this pull request.
> 
> 
> 
> >  +		 */
>  +		launch = (struct cpulaunch *)CKSEG0ADDR(CPULAUNCH);
> -+		launch += 2; /* MT7621 has 2 VPEs per core */
> -+		if (!(launch->flags & LAUNCH_FREADY))
> ++		launch++; /* core0:1 */
> ++		flags01 = launch->flags;
> ++		launch++; /* core1:0 */
> ++		if ((flags01 & LAUNCH_FREADY) && !(launch->flags & LAUNCH_FREADY))
> 
> @mrkiko Why you check `SYSC_REG_CHIP_NAME0` register instead of `SYSC_REG_CHIP_REV`?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/17834#discussion_r1946582812
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

namiltd commented:
So the test code should look like this:
```
		testval = _mt7621_get_soc_revq();
		testcores = FIELD_GET(CPU_ID, testval);
		pr_info("Using readl, it looks like my system has %u cores (testval = %u).\n",testcores,testval);

```

namiltd commented:
Try this: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18710778/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)

namiltd commented:
Try this: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18710778/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)

namiltd commented:
Try this: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18710834/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)

mrkiko commented:
I get the idea and I think it should work. But the patch doesn't apply for me at the moment.


On Fri, Feb 07, 2025 at 10:33:42AM -0800, Mieczyslaw Nalewaj wrote:
> Try this: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18710778/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)
> 
> 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/17834#issuecomment-2643700108
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

namiltd commented:
Fixed: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18711255/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)

mrkiko commented:
Doesn't apply for me. Do you mind if we meet on IRC to understand why?

On Fri, Feb 07, 2025 at 10:53:44AM -0800, Mieczyslaw Nalewaj wrote:
> Fixed: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18711255/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)
> 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/17834#issuecomment-2643732175
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

namiltd commented:
Please try downloading the patch again, I have corrected the file.

mrkiko commented:
so, downloading the file again I get:
--- a/arch/mips/include/asm/mips-cps.h
+++ b/arch/mips/include/asm/mips-cps.h
@@ -162,12 +162,15 @@ static inline uint64_t mips_cps_cluster_
  */
 static inline unsigned int mips_cps_numcores(unsigned int cluster)
 {
 	if (!mips_cm_present())
 		return 0;

+	if (IS_ENABLED(CONFIG_SOC_MT7621))
+		return 1 + FIELD_GET(0x20000, _mt7621_get_soc_revq());
+ 
 	/* Add one before masking to handle 0xff indicating no cores */
 	return FIELD_GET(CM_GCR_CONFIG_PCORES,
 			 mips_cps_cluster_config(cluster) + 1);

which doesn't apply. Trying to do it by hand.

On Fri, Feb 07, 2025 at 11:09:39AM -0800, Mieczyslaw Nalewaj wrote:
> Please try downloading the patch again, I have corrected the file.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/17834#issuecomment-2643783750
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

namiltd commented:
Please wait a moment, the file is being updated.

namiltd commented:
Fixed: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18711915/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)

namiltd commented:
Successfully tested on Valokuitu Plus EX400 DNA, two cores were detected correctly.

mrkiko commented:
It works great on R6220.
Feel free to add my tested-by tag.

On Fri, Feb 07, 2025 at 12:03:05PM -0800, Mieczyslaw Nalewaj wrote:
> Fixed: [320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch](https://github.com/user-attachments/files/18711915/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch)
> 
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/17834#issuecomment-2644015696
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

namiltd commented:
For tested-by I need name, surname and email.

msandber commented:
> For tested-by I need name, surname and email.

Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>

hauke commented on target/linux/ramips/patches-6.6/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch:
```diff
@@ -26,50 +26,36 @@ Suggested-by: Felix Fietkau <nbd@nbd.name>
  	if (!mips_cm_present())
  		return 0;
  
++	if (IS_ENABLED(CONFIG_SOC_MT7621))
++		return 1 + FIELD_GET(0x20000, __raw_readl(MT7621_SYSC_BASE + SYSC_REG_CHIP_REV));
++ 
```
The SYSC_REG_CHIP_REV is probably a fuse register of the SoC. Every SoC has such registers which allow the manufacture to configure the SoC. They are normally set after the testing was done. Some are even read by the hardware and then the hardware will deactivate some functionality based on which bit is set. This is often used to sell the same product to different markets for different prices or to mark products which achieved better clock rates in the tests.

I do not think it contains the number of cores, but if this is the low cost version or not. 

I prefer the original code from @DragonBluep much more.
```
	#define CHIP_REV_CORE_MASK		0x1
	#define CHIP_REV_CORE_SHIFT		17
	if (IS_ENABLED(CONFIG_SOC_MT7621)) {
		u32 sysc = __raw_readl(MT7621_SYSC_BASE + SYSC_REG_CHIP_REV);
		if (!((sysc >> CHIP_REV_CORE_SHIFT) & CHIP_REV_CORE_MASK))
			ncores = 1;
	}
```
This version checks if the bit is not set and then sets the ncores to 1 or you can return 1. I would not use the bit to do some calculation. 
If the MTK SDK uses the same code it should be safe.

namiltd replied:
It does not contain the number of cores but the version, it is just one bit equal to 0 (mt7621s) or 1 (mt7621a). Adding 1 to this we have the number of cores.

namiltd replied:
My code works exactly the same as the original code from @DragonBluep, but it's simpler.

Alternatively:
```
	if (IS_ENABLED(CONFIG_SOC_MT7621))
		return FIELD_GET(0x20000, __raw_readl(MT7621_SYSC_BASE + SYSC_REG_CHIP_REV)) ? 2 : 1;
```
or
```
	if (IS_ENABLED(CONFIG_SOC_MT7621))
		return (__raw_readl(MT7621_SYSC_BASE + SYSC_REG_CHIP_REV) & 0x20000) ? 2 : 1;
```

hauke replied:
I prefer the first alternative solution.

namiltd commented:
Done

DragonBluep commented on target/linux/ramips/patches-6.6/320-MIPS-add-support-for-buggy-MT7621S-core-detection.patch:
```diff
@@ -26,50 +26,36 @@ Suggested-by: Felix Fietkau <nbd@nbd.name>
  		return 0;
  
++	if (IS_ENABLED(CONFIG_SOC_MT7621))
++		return FIELD_GET(0x20000, __raw_readl(MT7621_SYSC_BASE + SYSC_REG_CHIP_REV)) ? 2 : 1;
```
Please don't touch the core detect logic of the MT7621**A**. Just return `1` for MT7621**S** here.

DragonBluep replied:
MT7621**A** can operate in single core mode. Ref: https://forum.openwrt.org/t/how-to-enable-hyperthreading-on-mediatek-mt7621at/78419

namiltd replied:
Done

namiltd replied:
Done

namiltd replied:
Done

namiltd committed:
ramips: improve MT7621S core detection

The proposed detection method was based on reading the LAUNCH_FREADY core flag.
However, this method only works before the cores are launched.
For this reason, the core number detection method has been changed to a simpler one.
For mt6721s the 17th revision bit is zero, hence we know that it is this chip,
so the number of cores is 1.

Fixes: https://github.com/openwrt/openwrt/issues/17764
Tested-by: Enrico Mioso <mrkiko.rs@gmail.com>
Tested-by: Simon Etzlstorfer <simon@etzi.at>
Tested-by: Mauri Sandberg <maukka@ext.kapsi.fi>
Co-authored-by: Shiji Yang <yangshiji66@qq.com>
Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/17834
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented:
Please approve this patxh for branch 24.10 as well.
